allennlp/data/dataset_readers/language_modeling.py

```diff
@@ -1,97 +1,160 @@
+        These are prepended to each line read by the dataset reader.
+    end_tokens : ``List[str]``, optional (default=``["</S>"]``)
+        These are appended to each line read by the dataset reader.
     """
     def __init__(self,
-                 tokens_per_instance: int = None,
+                 batch_size: int = 20,
+                 truncated_bptt_size: int = 35,
+                 fuzz_truncated_bptt_size: bool = True,
+                 bidirectional: bool = False,
                  tokenizer: Tokenizer = None,
                  token_indexers: Dict[str, TokenIndexer] = None,
-                 lazy: bool = False) -> None:
-        super().__init__(lazy)
+                 start_tokens: List[str] = None,
+                 end_tokens: List[str] = ["</S>"]) -> None:
+        super().__init__(lazy=False)
+        self._batch_size = batch_size
+        if truncated_bptt_size < 2:
+            raise ConfigurationError("truncated_bptt_size cannot be less than 2.")
+        self._truncated_bptt_size = truncated_bptt_size
+        self._fuzz_truncated_bptt_size = fuzz_truncated_bptt_size
+        self._bidirectional = bidirectional
         self._tokenizer = tokenizer or WordTokenizer()
         self._token_indexers = token_indexers or {"tokens": SingleIdTokenIndexer()}
-        self._tokens_per_instance = tokens_per_instance
 
-        # No matter how you want to represent the input, we'll always represent the output as a
-        # single token id.  This code lets you learn a language model that concatenates word
-        # embeddings with character-level encoders, in order to predict the word token that comes
-        # next.
-        self._output_indexer: Dict[str, TokenIndexer] = None
-        for name, indexer in self._token_indexers.items():
-            if isinstance(indexer, SingleIdTokenIndexer):
-                self._output_indexer = {name: indexer}
-                break
-        else:
-            self._output_indexer = {"tokens": SingleIdTokenIndexer()}
+        self._start_tokens = [Token(st) for st in (start_tokens or [])]
+        self._end_tokens = [Token(et) for et in (end_tokens or [])]
+        # Cache the batched tokens we read from data files.
+        self._all_batched_file_tokens = {}
 
     @overrides
     def _read(self, file_path: str):
-        # if `file_path` is a URL, redirect to the cache
-        file_path = cached_path(file_path)
+        if file_path not in self._all_batched_file_tokens:
+            logger.info('Loading data from %s', file_path)
+            # if `file_path` is a URL, redirect to the cache
+            file_path = cached_path(file_path)
 
-        with open(file_path, "r") as text_file:
-            instance_strings = text_file.readlines()
+            # Read the contents of the file into one long list of tokens,
+            # adding start and/or end tokens as necessary.
+            file_tokens = []
+            file_tokens.extend(self._start_tokens)
```

Comment:
Shouldn't this be in the forloop?

Reply:
Yup, you're correct (sorry I haven't looked at this PR in awhile). In general, though, language modeling of contiguous text doesn't use any start tokens and only uses eos tokens (e.g., see https://github.com/salesforce/awd-lstm-lm/blob/master/data.py#L34-L54 )

Reply:
yeah, although AFAIK ELMo actually used start tokens (https://github.com/allenai/allennlp/blob/master/tutorials/how_to/elmo.md#notes-on-statefulness-and-non-determinism)

Reply:
Right, but this dataset reader is not for training a LM like ELMo. ELMo is trained on shuffled sentences, hence why it needs a start token for each sentence. This dataset reader is for _contiguous text_ (e.g., books), like the corpora that the OpenAI GPT was trained on / the PTB LM benchmark most folks use. In this setting, people don't typically use start tokens.

Reply:
Ah, got it, thanks -- I assumed that stateful (RNN) language models were _always_ trained in contiguous fashion.
The shuffling bit might help generalization performance for sentence-level tasks (where the model doesn't learn to rely on an initialized hidden state), but I'm wondering how the same model with and without contiguous training would perform in transfer learning. 
Thanks!